File: src/com/esielkar/proxy/Proxy.java
package com.esielkar.proxy;

import java.util.Random;

public class Proxy implements ServiceInterface {
    private final Service realService;

    public Proxy(Service service) {
        realService = service;
    }

    private boolean checkAccess() {
        return new Random().nextInt() % 2 == 0;

    }

    @Override
    public void operation() {
        System.out.println("Proxy: operation");
        if (checkAccess()) {
            realService.operation();
        }
    }

}

File: dart/design_patterns/bin/proxy/proxy.dart
import 'dart:math';

import 'service.dart';

class Proxy implements ServiceInterface {
  final Service _realService;

  Proxy(Service service) : _realService = service;

  bool _checkAccess() {
    return Random().nextInt(2) % 2 == 0;
  }

  @override
  void operation() {
    print("Proxy: operation");
    if (_checkAccess()) {
      _realService.operation();
    }
  }
}
